orca-core/src/main/groovy/com/netflix/spinnaker/orca/pipeline/model/Execution.java

```diff
@@ -294,6 +294,15 @@ public Stage stageByRef(String refId) {
       .orElseThrow(() -> new IllegalArgumentException(String.format("No stage with refId %s exists", refId)));
   }
 
+  @Nonnull
+  public Stage parentStageById(String stageId) {
+    Stage stage = stageById(stageId);
+    while (stage.getParentStageId() != null) {
+      stage = stageById(stage.getParentStageId());
+    }
+    return stage;
+  }
```

Comment:
Couldn't this be a method on `Stage`? Then you don't need the parameter.

Reply:
And I just saw that Stage.getTopLevelStage() already does this, somehow missed that yesterday.